Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -153,567 +131,429 @@ final class ConnectionPool {
+    /// Called from `HTTP1ConnectionProvider.close` when client is shutting down.
+    func close() -> EventLoopFuture<Void> {
+        return self.channel.close()
     }
 
-    /// A connection provider of `HTTP/1.1` connections with a given `Key` (host, scheme, port)
-    ///
-    /// On top of enabling connection reuse this provider it also facilitates the creation
-    /// of concurrent requests as it has built-in politeness regarding the maximum number
-    /// of concurrent requests to the server.
-    class HTTP1ConnectionProvider: CustomStringConvertible {
-        /// The default `EventLoop` for this provider
-        ///
-        /// The default event loop is used to create futures and is used
-        /// when creating `Channel`s for requests for which the
-        /// `EventLoopPreference` is set to `.indifferent`
-        let eventLoop: EventLoop
-
-        /// The client configuration used to bootstrap new requests
-        private let configuration: HTTPClient.Configuration
-
-        /// The key associated with this provider
-        private let key: ConnectionPool.Key
-
-        /// The `State` of this provider
-        ///
-        /// This property holds data structures representing the current state of the provider
-        /// - Warning: This type isn't thread safe and should be accessed with proper
-        /// synchronization (see the `stateLock` property)
-        fileprivate var state: State
-
-        /// The lock used to access and modify the `state` property
-        ///
-        /// - Warning: This lock should always be acquired *after* `ConnectionPool`s `connectionProvidersLock` if used in combination with it.
-        fileprivate let stateLock = Lock()
-
-        /// The maximum number of concurrent connections to a given (host, scheme, port)
-        private let maximumConcurrentConnections: Int = 8
-
-        /// The pool this provider belongs to
-        private let parentPool: ConnectionPool
-
-        /// Creates a new `HTTP1ConnectionProvider`
-        ///
-        /// - parameters:
-        ///     - key: The `Key` (host, scheme, port) this provider is associated to
-        ///     - configuration: The client configuration used globally by all requests
-        ///     - initialConnection: The initial connection the pool initializes this provider with
-        ///     - parentPool: The pool this provider belongs to
-        init(key: ConnectionPool.Key, eventLoop: EventLoop, configuration: HTTPClient.Configuration, parentPool: ConnectionPool) {
-            self.eventLoop = eventLoop
-            self.configuration = configuration
-            self.key = key
-            self.parentPool = parentPool
-            self.state = State(eventLoop: eventLoop, parentPool: parentPool, key: key)
+    /// Sets idle timeout handler and channel inactivity listener.
+    func setIdleTimeout(timeout: TimeAmount?) {
+        _ = self.channel.pipeline.addHandler(IdleStateHandler(writeTimeout: timeout)).flatMap { _ in
```

Comment:
Idle handlers should usually go first in the pipeline.

Reply:
Even if the timeout handling handler is added immediately after it? I don't think there are any handlers that can (or should) handle timeouts in the pipeline at this point...

Reply:
The reason they should go first in the pipeline is so that they observe the TCP writes/reads, rather than something more structured that may be delayed. Otherwise slow connections risk being killed for being idle, when in practice there's data flow occurring and just not being seen due to intermediate buffering.

This function is potentially callable at any time, so we'll risk fewer bugs by explicitly calling out that the idle state handler should go first. I'm fine for the actual timeout handler to go wherever.

Reply:
Ah! That makes sense, done